components/script/dom/intersectionobserver.rs

```diff
@@ -164,42 +191,65 @@ impl IntersectionObserverMethods<crate::DomTypeHolder> for IntersectionObserver
+        // >    set this’s internal [[rootMargin]] slot to that. Otherwise, throw a SyntaxError exception.
+        let root_margin = if let Ok(margin) = parse_a_margin(init.rootMargin.as_ref()) {
+            margin
+        } else {
+            return Err(Error::Syntax)
+        };
 
-#[derive(JSTraceable, MallocSizeOf)]
-struct MarginRect {
-    top: LengthOrPercentage,
-    right: LengthOrPercentage,
-    bottom: LengthOrPercentage,
-    left: LengthOrPercentage,
-}
+        // > 4. Attempt to parse a margin from options.scrollMargin. If a list is returned,
+        // >    set this’s internal [[scrollMargin]] slot to that. Otherwise, throw a SyntaxError exception.
+        let scroll_margin = if let Ok(margin) = parse_a_margin(init.scrollMargin.as_ref()) {
+            margin
+        } else {
+            return Err(Error::Syntax)
+        };
 
-impl MarginRect {
-    fn zero() -> Self {
-        Self {
-            top: LengthOrPercentage::zero(),
-            right: LengthOrPercentage::zero(),
-            bottom: LengthOrPercentage::zero(),
-            left: LengthOrPercentage::zero(),
+        // > 5. Let thresholds be a list equal to options.threshold.
+        //
+        // Non-sequence value should be converted into Vec.
+        // Empty vec also generated for default value of threshold.
+        let mut threshold = match &init.threshold {
+            Some(DoubleOrDoubleSequence::Double(num)) => vec![*num],
+            Some(DoubleOrDoubleSequence::DoubleSequence(sequence)) => sequence.clone(),
+            None => Default::default(),
+        };
+
+        // > 6. If any value in thresholds is less than 0.0 or greater than 1.0, throw a RangeError exception.
+        threshold.iter().try_for_each(|num| {
+            match **num < 0.0 || **num > 1.0 {
+                true => Err(Error::Range("Value in thresholds should not be less than 0.0 or greater than 1.0".to_owned())),
+                false => Ok(()),
+            }
+        })?;
+
+        // > 7. Sort thresholds in ascending order.
+        threshold.sort_by(|lhs, rhs| lhs.partial_cmp(&**rhs).unwrap());
+
+
+        // > 8. If thresholds is empty, append 0 to thresholds.
+        if threshold.is_empty() {
+            threshold.push(Finite::wrap(0.));
         }
-    }
-}
 
-#[derive(JSTraceable, MallocSizeOf)]
-enum LengthOrPercentage {
-    // Length will be converted into pixel size on construction.
-    AbsolutePixelLength(f32),
-    Percentage(f32),
-}
+        // > 10. Let delay be the value of options.delay.
+        //
+        // Default value of delay is "0".
+        let mut delay = init.delay.unwrap_or(0);
 
-impl LengthOrPercentage {
-    const fn zero() -> Self {
-        Self::AbsolutePixelLength(0.0)
+        // > 11. If options.trackVisibility is true and delay is less than 100, set delay to 100.
+        if init.trackVisibility {
+            delay = delay.min(100);
```

Comment:
I believe we need `delay.max(100)` here, otherwise we'll take the lower value.

Reply:
Yes, thanks! My bad for the silly mistake.